src/gauging.jl

```diff
@@ -0,0 +1,58 @@
+"""Modify the diagonal elements of an ITensor with a specified function"""
+function modify_diagonal_els_itensor(A::ITensor, f::Function)
+  out = copy(A)
+  is = [[j for i in 1:length(inds(A))] for j in 1:mindim(out)]
+  for i in is
+    out[i...] = f(out[i...])
+  end
+
+  return out
+end
+
+"""Put an ITensorNetwork into the symmetric gauge and also return the message tensors (which are the diagonal bond matrices)"""
+function symmetrise_itensornetwork(ψ::ITensorNetwork)
+  ψψ = ψ ⊗ prime(dag(ψ); sites=[])
+  vertex_groups = nested_graph_leaf_vertices(partition(ψψ, group(v -> v[1], vertices(ψψ))))
+  mts = compute_message_tensors(ψψ; vertex_groups=vertex_groups)
+
+  ψsymm = copy(ψ)
+  symm_mts = copy(mts)
+
+  for e in edges(ψ)
+    vsrc, vdst = src(e), dst(e)
+
+    s1, s2 = find_subgraph((vsrc, 1), mts), find_subgraph((vdst, 1), mts)
+    edge_ind = commoninds(mts[s1 => s2], ψsymm[vsrc])
+    edge_ind_sim = sim(edge_ind)
+
+    X_D, X_U = eigen(mts[s1 => s2]; ishermitian=true)
+    Y_D, Y_U = eigen(mts[s2 => s1]; ishermitian=true)
```

Comment:
Maybe to handle some of the potential zeros, we could add a cutoff here. For example, we could default to `cutoff=10eps(real(scalartype(ψ)))` and control it with a kwarg in the top level function, for example:
```julia
function symmetric_gauge(
  ψ::ITensorNetwork;
  eigen_message_tensor_cutoff=10eps(real(scalartype(ψ))),
)
  # ...
  X_D, X_U = eigen(mts[s1 => s2]; ishermitian=true, cutoff=eigen_message_tensor_cutoff)
  # ...
end
```

Reply:
Okay yeah this sounds like a good idea. Will do that.